### PR TITLE
Allow default exporter with additional Quarkiverse exporters

### PR DESCRIPTION
**Pull request by amalsebs**
Fixes
-  Allow default exporter with additional Quarkiverse exporters #36500

### CODE REVIEW

File: integration-tests/opentelemetry/src/main/resources/application.properties [hunk 1]
```diff
@@ -22,3 +22,6 @@ quarkus.security.users.embedded.enabled=true
 quarkus.http.auth.basic=true
 
 quarkus.otel.traces.suppress-application-uris=/suppress-app-uri
+
+%test.quarkus.otel.experimental.otlp.default.enable=true
```
Comment (brunobat) [inline integration-tests/opentelemetry/src/main/resources/application.properties]: Can you please create a config profile just to the `%test.quarkus.otel.experimental.otlp.default.enable=true` in the test using it?
This shouldn't be affecting the other tests.

Comment (amalsebs) [reply integration-tests/opentelemetry/src/main/resources/application.properties]: Done

File: extensions/opentelemetry/deployment/src/test/java/io/quarkus/opentelemetry/deployment/exporter/otlp/OtlpDefaultBehaviorTest.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+import io.opentelemetry.sdk.metrics.export.MetricExporter;
+import io.quarkus.opentelemetry.runtime.exporter.otlp.tracing.LateBoundSpanProcessor;
+import io.quarkus.test.QuarkusExtensionTest;
+
+public class OtlpDefaultBehaviorTest {
+
+    @RegisterExtension
+    static final QuarkusExtensionTest config = new QuarkusExtensionTest()
+            .withApplicationRoot(root -> root.addClasses(MockExporterProducer.class))
+            .overrideConfigKey("quarkus.otel.experimental.otlp.default.enable", "false")
```
Comment (brunobat) [inline extensions/opentelemetry/deployment/src/test/java/io/quarkus/opentelemetry/deployment/exporter/otlp/OtlpDefaultBehaviorTest.java]: If it's the default behaviour, you shouldn't set the property in the test.

File: extensions/opentelemetry/deployment/src/test/java/io/quarkus/opentelemetry/deployment/exporter/otlp/OtlpDefaultBehaviorTest.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+            .overrideConfigKey("quarkus.otel.traces.exporter", "cdi")
+            .overrideConfigKey("quarkus.otel.metrics.exporter", "cdi")
+            .overrideConfigKey("quarkus.otel.logs.exporter", "cdi")
```
Comment (brunobat) [inline extensions/opentelemetry/deployment/src/test/java/io/quarkus/opentelemetry/deployment/exporter/otlp/OtlpDefaultBehaviorTest.java]: Same for CDI

File: extensions/opentelemetry/deployment/src/test/java/io/quarkus/opentelemetry/deployment/exporter/otlp/OtlpDefaultBehaviorTest.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+        for (Object ignored : instance) {
+            count++;
+        }
+        return count;
+    }
+
+    public static class MockExporterProducer {
+        @Produces
+        @Singleton
+        @jakarta.enterprise.inject.Typed(io.opentelemetry.sdk.trace.export.SpanExporter.class)
```
Comment (brunobat) [inline extensions/opentelemetry/deployment/src/test/java/io/quarkus/opentelemetry/deployment/exporter/otlp/OtlpDefaultBehaviorTest.java]: Please use imports

File: integration-tests/opentelemetry/src/test/java/io/quarkus/it/opentelemetry/util/CustomExporter.java [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+import java.util.Collection;
+
+import jakarta.enterprise.context.ApplicationScoped;
+
+import io.opentelemetry.sdk.common.CompletableResultCode;
+import io.opentelemetry.sdk.trace.data.SpanData;
+import io.opentelemetry.sdk.trace.export.SpanExporter;
+
+@ApplicationScoped
+public class CustomExporter implements SpanExporter {
```
Comment (brunobat) [inline integration-tests/opentelemetry/src/test/java/io/quarkus/it/opentelemetry/util/CustomExporter.java]: This shouldn't be on main but instantiated in one of the test classes. This is interfering with the other tests.

File: extensions/opentelemetry/deployment/src/main/java/io/quarkus/opentelemetry/deployment/exporter/otlp/OtlpExporterProcessor.java [hunk 6]
```diff
@@ -140,11 +140,20 @@ void errorIfOtlpUpstreamIsUsed(CombinedIndexBuildItem indexBuildItem,
...
-            // if there is an external exporter, we don't want to create the default one
+
+        if (!externalOtelExporterBuildItem.isEmpty() && !otelBuildConfig.defaultExporterEnabled()) {
+            // unless explicitly configured we don't want to create the default one if there is an external exporter.
+            // External exporter also use synthetic beans. However, synthetic beans don't show in the BeanDiscoveryFinishedBuildItem
+            return;
+        }
+
+        if (!beanDiscovery.beanStream().withBeanType(DotName.createSimple(SpanExporter.class.getName())).isEmpty()
+                && !otelBuildConfig.defaultExporterEnabled()) {
```
Comment (brunobat) [inline extensions/opentelemetry/deployment/src/main/java/io/quarkus/opentelemetry/deployment/exporter/otlp/OtlpExporterProcessor.java]: I'm not sure I understand this. If we find a CDI `SpanExporter` implementation you are not including it... This doesn't seem right.

File: integration-tests/opentelemetry/src/test/java/io/quarkus/it/opentelemetry/OtlpCoexistenceIT.java [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+                .get("/export/count")
+                .then()
+                .statusCode(200)
+                .extract().asString();
+
+        assertTrue(response.contains("InMemorySpanExporter"),
+                "InMemory exporter should be registered");
+
+        String[] exporters = response.split(",");
+        assertTrue(exporters.length >= 2,
```
Comment (brunobat) [inline integration-tests/opentelemetry/src/test/java/io/quarkus/it/opentelemetry/OtlpCoexistenceIT.java]: I think that even without your code this test would pass because the `LateBoundSpanProcessor` also implements `SpanProcessor`.

File: integration-tests/opentelemetry/src/main/java/io/quarkus/it/opentelemetry/ExporterResource.java [hunk 8]
```diff
@@ -101,6 +118,8 @@ private static boolean isPathFound(String path, Attributes attributes) {
     static class InMemorySpanExporterProducer {
         @Produces
         @Singleton
+        @Priority(1)
+        @Alternative
```
Comment (brunobat) [inline integration-tests/opentelemetry/src/main/java/io/quarkus/it/opentelemetry/ExporterResource.java]: Please don't do this. As it affects all other existing tests.

File: extensions/opentelemetry/deployment/src/test/java/io/quarkus/opentelemetry/deployment/exporter/otlp/OtlpCoexistenceTest.java [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+    }
+
+    @Test
+    void testLogsCoexist() {
+        assertTrue(logRecordExporter.isResolvable(),
+                "Logs: LogRecordExporter should be resolvable");
+        assertFalse(logRecordExporter.get() instanceof NoopLogRecordExporter,
+                "Logs: Should not be a NoopLogRecordExporter");
+    }
+}
```
Comment (brunobat) [inline extensions/opentelemetry/deployment/src/test/java/io/quarkus/opentelemetry/deployment/exporter/otlp/OtlpCoexistenceTest.java]: Missing new line

**Review by brunobat — CHANGES_REQUESTED** [review]
Thanks for the PR @amalsebs but we need changes.

**Comment by gsmet** [comment]
@amalsebs :wave: do you plan to revisit this?

**Comment by amalsebs** [comment]
> @amalsebs 👋 do you plan to revisit this?

yes

File: extensions/opentelemetry/deployment/src/test/java/io/quarkus/opentelemetry/deployment/exporter/otlp/OtlpDefaultBehaviorTest.java (hunk 2 above)
Comment (amalsebs) [reply extensions/opentelemetry/deployment/src/test/java/io/quarkus/opentelemetry/deployment/exporter/otlp/OtlpDefaultBehaviorTest.java]: Removed the manual property overrides for default behaviour.

File: extensions/opentelemetry/deployment/src/test/java/io/quarkus/opentelemetry/deployment/exporter/otlp/OtlpDefaultBehaviorTest.java (hunk 3 above)
Comment (amalsebs) [reply extensions/opentelemetry/deployment/src/test/java/io/quarkus/opentelemetry/deployment/exporter/otlp/OtlpDefaultBehaviorTest.java]: Removed the manual property overrides and CDI forced types.

File: extensions/opentelemetry/deployment/src/test/java/io/quarkus/opentelemetry/deployment/exporter/otlp/OtlpDefaultBehaviorTest.java (hunk 4 above)
Comment (amalsebs) [reply extensions/opentelemetry/deployment/src/test/java/io/quarkus/opentelemetry/deployment/exporter/otlp/OtlpDefaultBehaviorTest.java]: Done

File: integration-tests/opentelemetry/src/test/java/io/quarkus/it/opentelemetry/util/CustomExporter.java (hunk 5 above)
Comment (amalsebs) [reply integration-tests/opentelemetry/src/test/java/io/quarkus/it/opentelemetry/util/CustomExporter.java]: Moved CustomExporter to src/test and removed @ApplicationScoped to prevent interference with other ITs.

File: extensions/opentelemetry/deployment/src/main/java/io/quarkus/opentelemetry/deployment/exporter/otlp/OtlpExporterProcessor.java (hunk 6 above)
Comment (amalsebs) [reply extensions/opentelemetry/deployment/src/main/java/io/quarkus/opentelemetry/deployment/exporter/otlp/OtlpExporterProcessor.java]: Refactored to be clearer. The return only skips the synthetic OTLP bean, it doesn't touch the discovered CDI beans. This allows the opt-in coexistence without breaking the standard back-off.

File: integration-tests/opentelemetry/src/test/java/io/quarkus/it/opentelemetry/OtlpCoexistenceIT.java (hunk 7 above)
Comment (amalsebs) [reply integration-tests/opentelemetry/src/test/java/io/quarkus/it/opentelemetry/OtlpCoexistenceIT.java]: Updated the assertion. The test now explicitly verifies that LateBoundSpanProcessor and BatchSpanProcessor are are active simultaneously.

File: extensions/opentelemetry/deployment/src/test/java/io/quarkus/opentelemetry/deployment/exporter/otlp/OtlpCoexistenceTest.java (hunk 9 above)
Comment (amalsebs) [reply extensions/opentelemetry/deployment/src/test/java/io/quarkus/opentelemetry/deployment/exporter/otlp/OtlpCoexistenceTest.java]: Added new line

File: integration-tests/opentelemetry/src/main/java/io/quarkus/it/opentelemetry/ExporterResource.java (hunk 8 above)
Comment (amalsebs) [reply integration-tests/opentelemetry/src/main/java/io/quarkus/it/opentelemetry/ExporterResource.java]: Removed

File: integration-tests/opentelemetry/src/test/java/io/quarkus/it/opentelemetry/OtlpCoexistenceIT.java [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+import io.quarkus.test.junit.QuarkusTestProfile;
+import io.quarkus.test.junit.TestProfile;
+
+@QuarkusTest
+@TestProfile(OtlpCoexistenceIT.OtlpEnabledProfile.class)
+public class OtlpCoexistenceIT {
+
+    @Produces
+    @Singleton
+    CustomExporter customExporter = new CustomExporter();
```
Comment (brunobat) [inline integration-tests/opentelemetry/src/test/java/io/quarkus/it/opentelemetry/OtlpCoexistenceIT.java]: In order to understand if the CDI infection is working, shouldn't the `CustomExporter` be a CDI bean as well?

File: integration-tests/opentelemetry/src/test/java/io/quarkus/it/opentelemetry/OtlpCoexistenceIT.java [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+
+        assertTrue(response.contains("CustomExporter"),
+                "CustomExporter exporter should be registered");
+
+        assertTrue(response.contains("LateBoundSpanProcessor"), "Default OTLP processor missing");
+    }
+
+    public static class OtlpEnabledProfile implements QuarkusTestProfile {
+        @Override
+        public Map<String, String> getConfigOverrides() {
```
Comment (brunobat) [inline integration-tests/opentelemetry/src/test/java/io/quarkus/it/opentelemetry/OtlpCoexistenceIT.java]: What will happen when have metrics and logs on? Do we get an ambiguous resolution exception?

File: extensions/opentelemetry/deployment/src/main/java/io/quarkus/opentelemetry/deployment/exporter/otlp/OtlpExporterProcessor.java [hunk 12]
```diff
@@ -140,11 +141,18 @@ void errorIfOtlpUpstreamIsUsed(CombinedIndexBuildItem indexBuildItem,
...
+            BeanDiscoveryFinishedBuildItem beanDiscovery,
             CoreVertxBuildItem vertxBuildItem,
             List<ExternalOtelExporterBuildItem> externalOtelExporterBuildItem,
             BuildProducer<SyntheticBeanBuildItem> syntheticBeanBuildItemBuildProducer) {
-        if (!externalOtelExporterBuildItem.isEmpty()) {
-            // if there is an external exporter, we don't want to create the default one
+
+        // Check if any exporter already exists (External Synthetic or CDI Bean)
+        boolean externalExporterExists = !externalOtelExporterBuildItem.isEmpty() ||
+                !beanDiscovery.beanStream().withBeanType(SPAN_EXPORTER).isEmpty();
```
Comment (brunobat) [inline extensions/opentelemetry/deployment/src/main/java/io/quarkus/opentelemetry/deployment/exporter/otlp/OtlpExporterProcessor.java]: I continue not to agree with this. `externalOtelExporterBuildItems` should be all `SpanExporter` beans.
Also, SPI loaded exporters like the upstream OTLP ones will not show here.
We need more tests to reliably understand what will happen in these 3 cases:
* User provided CDI exporter
* Exporter extension using  `ExternalOtelExporterBuildItem`
* SPI loaded OTel OTLP exporter... the one using the OkHttp exporters.

File: extensions/opentelemetry/deployment/src/main/java/io/quarkus/opentelemetry/deployment/exporter/otlp/OtlpExporterProcessor.java [hunk 13]
```diff
@@ -52,7 +53,7 @@ static class OtlpTracingExporterEnabled implements BooleanSupplier {
         public boolean getAsBoolean() {
             return otelBuildConfig.enabled() &&
                     otelBuildConfig.traces().enabled().orElse(Boolean.TRUE) &&
-                    otelBuildConfig.traces().exporter().contains(CDI_VALUE) &&
+                    (otelBuildConfig.traces().exporter().contains(CDI_VALUE) || otelBuildConfig.defaultExporterEnabled()) &&
```
Comment (brunobat) [inline extensions/opentelemetry/deployment/src/main/java/io/quarkus/opentelemetry/deployment/exporter/otlp/OtlpExporterProcessor.java:56]: If a user sets `quarkus.otel.traces.exporter=none` but `defaultExporterEnabled=true`, the OTLP exporter build steps still run. That's not good.

File: extensions/opentelemetry/deployment/src/test/java/io/quarkus/opentelemetry/deployment/exporter/otlp/OtlpCoexistenceTest.java [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.opentelemetry.sdk.logs.export.LogRecordExporter;
+import io.opentelemetry.sdk.metrics.export.MetricExporter;
+import io.quarkus.opentelemetry.runtime.exporter.otlp.logs.NoopLogRecordExporter;
+import io.quarkus.opentelemetry.runtime.exporter.otlp.metrics.NoopMetricExporter;
+import io.quarkus.opentelemetry.runtime.exporter.otlp.tracing.LateBoundSpanProcessor;
+import io.quarkus.test.QuarkusExtensionTest;
+
+public class OtlpCoexistenceTest {
```
Comment (brunobat) [inline extensions/opentelemetry/deployment/src/test/java/io/quarkus/opentelemetry/deployment/exporter/otlp/OtlpCoexistenceTest.java:17]: You never added a custom exporter for this test

**Review by brunobat — CHANGES_REQUESTED** [review]

**Comment by brunobat** [comment]
We are missing a lot more tests for edge cases.

**Comment by cescoffier** [comment]
🤖 Automatically associated with the **WG - Observability.Next** Working Group based on AI classification.

**Comment by cescoffier** [comment]
@brunobat can you have another look?

File: integration-tests/opentelemetry/src/test/java/io/quarkus/it/opentelemetry/OtlpCoexistenceIT.java (hunk 10 above)
Comment (amalsebs) [reply integration-tests/opentelemetry/src/test/java/io/quarkus/it/opentelemetry/OtlpCoexistenceIT.java]: Fixed — it's now a proper @ApplicationScoped bean (CustomExporterAndProcessor) with @Produces methods, nested inside OtlpCoexistenceIT.

File: integration-tests/opentelemetry/src/test/java/io/quarkus/it/opentelemetry/OtlpCoexistenceIT.java (hunk 11 above)
Comment (amalsebs) [reply integration-tests/opentelemetry/src/test/java/io/quarkus/it/opentelemetry/OtlpCoexistenceIT.java]: Yes — this was a real bug. MetricsExporterCDIProvider/LogsExporterCDIProvider threw AmbiguousResolutionException when multiple exporter beans existed. Fixed with CompositeMetricExporter/CompositeLogRecordExporter, which fan out to all resolved beans instead of calling .get() on an ambiguous Instance. Covered by OtlpCoexistenceTest and OtlpMultipleCdiExportersWithoutDefaultTest.

File: extensions/opentelemetry/deployment/src/main/java/io/quarkus/opentelemetry/deployment/exporter/otlp/OtlpExporterProcessor.java (hunk 12 above)
Comment (amalsebs) [reply extensions/opentelemetry/deployment/src/main/java/io/quarkus/opentelemetry/deployment/exporter/otlp/OtlpExporterProcessor.java]: All three now have dedicated tests: user CDI exporter (OtlpCoexistenceTest, OtlpDefaultBehaviorTest), ExternalOtelExporterBuildItem (new OtlpExternalExporterBuildItemTest), and SPI loaded upstream OTLP presence (new OtlpUpstreamExporterPresentTest).

File: extensions/opentelemetry/deployment/src/main/java/io/quarkus/opentelemetry/deployment/exporter/otlp/OtlpExporterProcessor.java (hunk 13 above)
Comment (amalsebs) [reply extensions/opentelemetry/deployment/src/main/java/io/quarkus/opentelemetry/deployment/exporter/otlp/OtlpExporterProcessor.java:56]: Verified with a new test, OtlpExporterNoneDisablesDefaultTest  none correctly takes precedence over the coexistence flag for all three signals.

File: extensions/opentelemetry/deployment/src/test/java/io/quarkus/opentelemetry/deployment/exporter/otlp/OtlpCoexistenceTest.java (hunk 14 above)
Comment (amalsebs) [reply extensions/opentelemetry/deployment/src/test/java/io/quarkus/opentelemetry/deployment/exporter/otlp/OtlpCoexistenceTest.java:17]: OtlpCoexistenceTest now uses MockExporterProducer, a real custom exporter bean, added via the test's application root.

**Comment by amalsebs** [comment]
> We are missing a lot more tests for edge cases.

Added four new tests covering the gaps discussed above: exporter=none precedence, ExternalOtelExporterBuildItem, SPI-detected upstream OTLP, and multiple non-default CDI exporters coexisting without the flag. Happy to add more if there are specific scenarios still of concern.

**Comment by brunobat** [comment]
In my queue for review.

**Comment by cescoffier** [comment]
🤖 Automatically associated with the **WG - Observability.Next** Working Group based on AI classification.

File: integration-tests/opentelemetry/src/test/java/io/quarkus/it/opentelemetry/OtlpCoexistenceIT.java [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+import io.opentelemetry.sdk.metrics.export.MetricExporter;
+import io.opentelemetry.sdk.testing.exporter.InMemorySpanExporter;
+import io.opentelemetry.sdk.trace.export.SpanExporter;
+import io.quarkus.test.junit.QuarkusTest;
+import io.quarkus.test.junit.QuarkusTestProfile;
+import io.quarkus.test.junit.TestProfile;
+
+@QuarkusTest
+@TestProfile(OtlpCoexistenceIT.OtlpEnabledProfile.class)
+public class OtlpCoexistenceIT {
```
Comment (brunobat) [inline integration-tests/opentelemetry/src/test/java/io/quarkus/it/opentelemetry/OtlpCoexistenceIT.java]: Please remane to OtlpCoexistenceTest, the *IT sufix is reserved to native mode tests

Comment (brunobat) [reply integration-tests/opentelemetry/src/test/java/io/quarkus/it/opentelemetry/OtlpCoexistenceIT.java]: BTW, a native mode test is missing.

File: extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/config/build/OTelBuildConfig.java [hunk 16]
```diff
@@ -48,6 +49,22 @@ public interface OTelBuildConfig {
...
+     * By default, the default OTLP exporter is not created if another exporter
+     * (CDI bean or Quarkiverse extension) is already present for a given signal.
+     * If set to <code>true</code>, the default OTLP exporter is created regardless
+     * of other exporters present, as long as it is otherwise enabled
+     * (see {@code quarkus.otel.exporter.otlp.enabled} and the per-signal
+     * {@code exporter} properties).
+     * <p>
+     * Defaults to <code>false</code>.
+     */
+    @WithName("experimental.otlp.default.enable")
```
Comment (brunobat) [inline extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/config/build/OTelBuildConfig.java]: The nesting `experimental.otlp.default.enable` is unusual. `experimental.otlp-default-enable` or similar seems better.

File: extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/exporter/otlp/logs/CompositeLogRecordExporter.java [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+        if (delegates.size() == 1) {
+            return delegates.get(0);
+        }
+        return new CompositeLogRecordExporter(delegates);
+    }
+
+    @Override
+    public CompletableResultCode export(Collection<LogRecordData> logs) {
+        List<CompletableResultCode> results = delegates.stream()
+                .map(delegate -> delegate.export(logs))
```
Comment (brunobat) [inline extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/exporter/otlp/logs/CompositeLogRecordExporter.java]: Please avoid Lambdas in Extension Runtime Code. 

```
// Instead of:
List<CompletableResultCode> results = delegates.stream()
    .map(delegate -> delegate.export(logs))
    .toList();

// Use:
List<CompletableResultCode> results = new ArrayList<>(delegates.size());
for (LogRecordExporter delegate : delegates) {
    results.add(delegate.export(logs));
}
```

Same for the others bellow.

File: extensions/opentelemetry/deployment/src/main/java/io/quarkus/opentelemetry/deployment/exporter/otlp/OtlpExporterProcessor.java [hunk 18]
```diff
@@ -109,6 +112,11 @@ private boolean isTracesOtlp(OTelBuildConfig otelBuildConfig) {
         }
     }
 
+    private boolean isOtlpExporterPresent(CombinedIndexBuildItem indexBuildItem) {
```
Comment (brunobat) [inline extensions/opentelemetry/deployment/src/main/java/io/quarkus/opentelemetry/deployment/exporter/otlp/OtlpExporterProcessor.java]: `isOtlpExporterPresent` is unreliable. 
The method scans the Jandex CombinedIndex for classes under `io.opentelemetry.exporter.otlp.`, but the upstream opentelemetry-exporter-otlp JAR has no Jandex index or beans.xml. Its classes won't appear in the combined index unless the user explicitly adds quarkus.index-dependency for it. 
The `OtlpUpstreamExporterPresentTest` works only because it places a marker class in the application root, which IS always indexed, however a real user adding the upstream JAR as a Maven dependency would not trigger this detection.

File: extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/exporter/otlp/metrics/CompositeMetricExporter.java [hunk 19]
```diff
@@ -0,0 +1,66 @@
...
+            return delegates.get(0);
+        }
+        return new CompositeMetricExporter(delegates);
+    }
+
+    @Override
+    public AggregationTemporality getAggregationTemporality(InstrumentType instrumentType) {
+        // Use the first delegate's preference; OTel does not support per-exporter temporality
+        // negotiation for a single MeterProvider registration.
+        return delegates.get(0).getAggregationTemporality(instrumentType);
```
Comment (brunobat) [inline extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/exporter/otlp/metrics/CompositeMetricExporter.java:48]: This is a semantic correctness issue that should at minimum be logged as a warning, or
  documented as a known limitation.

File: extensions/opentelemetry/deployment/src/main/java/io/quarkus/opentelemetry/deployment/exporter/otlp/OtlpExporterProcessor.java [hunk 20]
```diff
@@ -145,17 +153,26 @@ UnremovableBeanBuildItem ensureSpanExporterIsRetained() {
...
+                !beanDiscovery.beanStream().withBeanType(SPAN_EXPORTER).isEmpty() ||
+                isOtlpExporterPresent(indexBuildItem);
+
+        // return only if an exporter exists AND the coexistence flag is NOT enabled
+        if (externalExporterExists && !otelBuildConfig.defaultExporterEnabled()) {
             return;
         }
-        if (!beanDiscovery.beanStream().withBeanType(SPAN_EXPORTER).isEmpty()) {
-            // if there is a SpanExporter bean impl around, we don't want to create the default one
+        if (!beanDiscovery.beanStream().withBeanType(SPAN_EXPORTER).isEmpty() && !otelBuildConfig.defaultExporterEnabled()) {
```
Comment (brunobat) [inline extensions/opentelemetry/deployment/src/main/java/io/quarkus/opentelemetry/deployment/exporter/otlp/OtlpExporterProcessor.java]: This might be dead code as it might have already returned for the tested condition.

File: integration-tests/opentelemetry/src/main/java/io/quarkus/it/opentelemetry/ExporterResource.java [hunk 21]
```diff
@@ -33,6 +34,8 @@ public class ExporterResource {
     InMemoryMetricExporter inMemoryMetricExporter;
     @Inject
     InMemoryLogRecordExporter inMemoryLogRecordExporter;
+    @Inject
+    OpenTelemetry openTelemetry;
```
Comment (brunobat) [inline integration-tests/opentelemetry/src/main/java/io/quarkus/it/opentelemetry/ExporterResource.java]: This is not used, why is it needed?

File: integration-tests/opentelemetry/src/test/java/io/quarkus/it/opentelemetry/util/CustomExporter.java [hunk 22]
```diff
@@ -0,0 +1,24 @@
+package io.quarkus.it.opentelemetry.util;
+
+import java.util.Collection;
+
+import io.opentelemetry.sdk.common.CompletableResultCode;
+import io.opentelemetry.sdk.trace.data.SpanData;
+import io.opentelemetry.sdk.trace.export.SpanExporter;
+
+public class CustomExporter implements SpanExporter {
```
Comment (brunobat) [inline integration-tests/opentelemetry/src/test/java/io/quarkus/it/opentelemetry/util/CustomExporter.java]: Were is this being used?

**Review by brunobat — CHANGES_REQUESTED** [review]
Sorry for taking so long to review this.
I had PTOs and this is not an easy feature... Thanks for being bold on taking it. 

I created an app running your code with both these extensions:

```xml
        <dependency>
            <groupId>io.quarkus</groupId>
            <artifactId>quarkus-micrometer-opentelemetry</artifactId>
        </dependency>
        <dependency>
            <groupId>io.quarkiverse.opentelemetry.exporter</groupId>
            <artifactId>quarkus-opentelemetry-exporter-azure</artifactId>
            <version>3.31.4.0</version>
        </dependency>
```
and the new config set to true:

```application.properties
quarkus.otel.experimental.otlp.default.enable=true
```

and in the logs I get:

```log
2026-08-19 12:55:52,432 WARN  [io.quarkus.config] (Quarkus Main Thread) Unrecognized configuration key "quarkus.otel.experimental.otlp.default.enable" was provided; it will be ignored; verify that the dependency extension for this configuration is set or that you did not make a typo

```

File: integration-tests/opentelemetry/src/test/java/io/quarkus/it/opentelemetry/OtlpCoexistenceIT.java (hunk 15 above)
Comment (amalsebs) [reply integration-tests/opentelemetry/src/test/java/io/quarkus/it/opentelemetry/OtlpCoexistenceIT.java]: Done, split into OtlpCoexistenceTest (the actual logic) and a thin OtlpCoexistenceIT that just extends it with @QuarkusIntegrationTest, so we get native coverage too now.

File: extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/config/build/OTelBuildConfig.java (hunk 16 above)
Comment (amalsebs) [reply extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/config/build/OTelBuildConfig.java]: Fixed, renamed to experimental.otlp-default-enable.

File: extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/exporter/otlp/logs/CompositeLogRecordExporter.java (hunk 17 above)
Comment (amalsebs) [reply extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/exporter/otlp/logs/CompositeLogRecordExporter.java]: Fixed in both CompositeMetricExporter and CompositeLogRecordExporter, replaced with for loops.

File: extensions/opentelemetry/deployment/src/main/java/io/quarkus/opentelemetry/deployment/exporter/otlp/OtlpExporterProcessor.java (hunk 18 above)
Comment (amalsebs) [reply extensions/opentelemetry/deployment/src/main/java/io/quarkus/opentelemetry/deployment/exporter/otlp/OtlpExporterProcessor.java]: Removed isOtlpExporterPresent entirely. Since your azure exporter goes through ExternalOtelExporterBuildItem, that's already covered separately.

File: extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/exporter/otlp/metrics/CompositeMetricExporter.java (hunk 19 above)
Comment (amalsebs) [reply extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/exporter/otlp/metrics/CompositeMetricExporter.java:48]: Documented in the class javadoc and inline comment.

File: extensions/opentelemetry/deployment/src/main/java/io/quarkus/opentelemetry/deployment/exporter/otlp/OtlpExporterProcessor.java (hunk 20 above)
Comment (amalsebs) [reply extensions/opentelemetry/deployment/src/main/java/io/quarkus/opentelemetry/deployment/exporter/otlp/OtlpExporterProcessor.java]: Removed, you were right it was redundant.

File: integration-tests/opentelemetry/src/main/java/io/quarkus/it/opentelemetry/ExporterResource.java (hunk 21 above)
Comment (amalsebs) [reply integration-tests/opentelemetry/src/main/java/io/quarkus/it/opentelemetry/ExporterResource.java]: Removed.

File: integration-tests/opentelemetry/src/test/java/io/quarkus/it/opentelemetry/util/CustomExporter.java (hunk 22 above)
Comment (amalsebs) [reply integration-tests/opentelemetry/src/test/java/io/quarkus/it/opentelemetry/util/CustomExporter.java]: Nowhere, deleted it, it was replaced by the nested bean in the test class a while back.

**Comment by amalsebs** [comment]
> Sorry for taking so long to review this. I had PTOs and this is not an easy feature... Thanks for being bold on taking it.
> 
> I created an app running your code with both these extensions:
> 
> ```
>         <dependency>
>             <groupId>io.quarkus</groupId>
>             <artifactId>quarkus-micrometer-opentelemetry</artifactId>
>         </dependency>
>         <dependency>
>             <groupId>io.quarkiverse.opentelemetry.exporter</groupId>
>             <artifactId>quarkus-opentelemetry-exporter-azure</artifactId>
>             <version>3.31.4.0</version>
>         </dependency>
> ```
> 
> and the new config set to true:
> 
> ```ini
> quarkus.otel.experimental.otlp.default.enable=true
> ```
> 
> and in the logs I get:
> 
> ```
> 2026-08-19 12:55:52,432 WARN  [io.quarkus.config] (Quarkus Main Thread) Unrecognized configuration key "quarkus.otel.experimental.otlp.default.enable" was provided; it will be ignored; verify that the dependency extension for this configuration is set or that you did not make a typo
> ```

Thank you Bruno for the detailed testing.
Renamed the config to experimental.otlp-default-enable. Would you mind re-testing with your azure setup when you get a chance, just want to confirm it's actually resolved for you and not just in our own tests.

**Comment by brunobat** [comment]
Sorry @amalsebs. I took a deep dive on the PR and I realize there are too many things to fix:

- The runtime guards are not needed because everything can be effectively controlled at build time. 
- `OpenTelemetryProcessor` changes are not needed, this can be done in the `OtlpExporterProcessor`.
- The property should live in the `OtlpExporterBuildConfig`.
- We do need to check the bean manager to check if implementations of the exporters do exist. User code might be adding them, like what we do in the tests. 
- We need more effective tests with the real exporter, in `integration-tests/opentelemetry-vertx-exporter`
- The `CompositeMetricExporter` has limitations that can only be overcome by a rewrite of the metrics exporter instantiation, out of scope for now. 

This is taking too long and I ended up creating a new PR. See: https://github.com/quarkusio/quarkus/pull/56423

I've added you as co-author. I hope you don't mind. 
Thanks for your work and I hope you can review the new PR.
